File: src/1-download-data.R
# 1-download-data.R
# ===========================================================================
# Download data for the earnings event study to local parquet files.
#
# This script downloads raw tables and merges them locally. It showcases three 
# different methods for downloading data from WRDS, each with different 
# RAM usage and speed characteristics:
#   1. Download CCM link + CRSP stocknames (small — simple collect)
#   2. Download Compustat fundq (medium — chunked ParquetFileWriter)
#   3. Download CRSP daily returns (large — chunked ParquetFileWriter)
#   4. Download market index returns (using download_wrds() wrapper)
#
# Each download demonstrates a different method, from simplest to most
# RAM-efficient.
#
#
# WHY THREE METHODS?
# The simplest approach (collect) loads the entire table into R memory.
# This works for small tables, but for large ones like CRSP daily (~100M rows)
# it can exceed your RAM and crash R. The chunked ParquetFileWriter method
# streams data in batches, so peak memory stays bounded regardless of table size.
#
# Benchmarks on comp.funda (935K rows x 1000 columns, 362 MB parquet):
#   collect():              528s,  18.4 GB RAM
#   Chunked ParquetWriter:  447s,   7.8 GB RAM
# The download_wrds() function in utils.R auto-sizes batches based on a
# target RAM limit (default 8 GB) so you don't have to think about it.
# ===========================================================================


# Setup ------------------------------------------------------------------------

# Check if the pacman package is installed, and install it if not.
if (!require("pacman")) install.packages("pacman")

# Load all the necessary packages using pacman::p_load. 
# This will install any missing packages automatically.
pacman::p_load(dotenv, keyring, dbplyr, RPostgres, DBI, glue, arrow,
               tictoc,
               tidyverse)

# Read from the .env file to get the data and output directories. 
#This allows us to keep our directory paths configurable for each user/machine
# and separate from the code.
load_dot_env(".env")

# reading the .env file created some environment variables 
# that we can access with Sys.getenv() to get the data and output directories.
data_dir   <- Sys.getenv("DATA_DIR")
output_dir <- Sys.getenv("OUTPUT_DIR")

# this loads a separate script called utils.R that contains helper functions. 
# This makes it easy to reuse utility functions and keeps our main script clean.
source("src/utils.R")


# Connect to WRDS --------------------------------------------------------------

# We use the keyring package to securely store your WRDS credentials in your
# operating system's credential store (Windows Credential Manager, macOS
# Keychain, etc.). This is more secure than putting passwords in .env files.
# Note: Do NOT put passwords in .env - use keyring for secrets.
#
# FIRST TIME SETUP 
# run these two lines in your R console if you did not do it in setup.R:
#   keyring::key_set("wrds_user")   # Will prompt for your WRDS username
#   keyring::key_set("wrds_pw")     # Will prompt for your WRDS password
# This stores credentials securely in your system's keychain.
# You only need to do this once per computer.
#
# To update stored credentials (e.g., after a password change), just re-run
# the key_set lines above.

wrds <- dbConnect(Postgres(),
                  host     = 'wrds-pgdata.wharton.upenn.edu',
                  port     = 9737,
                  user     = keyring::key_get("wrds_user"),
                  password = keyring::key_get("wrds_pw"),
                  sslmode  = 'require',
                  dbname   = 'wrds')

# If connection is successful, this should print connection info
wrds


# Download CCM link table ------------------------------------------------------

# The CCM link table maps Compustat gvkey to CRSP permno over time.
# It's small (~87K rows after filtering) so we just collect() it directly.
# collect() is the simplest download method — fine when the table fits in RAM.

# There are many ways to check how long a code takes.
# The tictoc package is a simple one. 
# tic() starts the timer, toc() stops it and prints the elapsed time.
tictoc::tic() #start timer
  
  # Download the full CCM link table — we'll filter it later when merging.
  # Keeping the raw file means you can reuse it with different link criteria.
  ccm_link <- tbl(wrds, in_schema("crsp", "ccmxpf_lnkhist")) |>
    collect()

tictoc::toc() #stop timer

# Check the result
nrow(ccm_link) # number of observations (rows)
head(ccm_link) # preview the first few rows to check the data looks correct

# Save the downloaded file to disk as a parquet file
write_parquet(ccm_link, glue("{data_dir}/ccm-link.parquet"))


# Download CRSP stocknames_v2 --------------------------------------------------

# This table will be used to collect SIC codes. 

# SIC codes come from CRSP rather than Compustat because:
#   - stocknames_v2 has time-varying SIC codes (matched by date range)
#   - It is current through 2025 (old stocknames table froze at 2024-12-31)
#   - We already have permno from the CCM link, so it's a natural join
# This is a small table, so collect() is fine.

tictoc::tic()

  stocknames <- tbl(wrds, in_schema("crsp", "stocknames_v2")) |>
    # example of first selecting only certain columns that we need to download
    select(permno, namedt, nameenddt, siccd) |>
    collect()

tictoc::toc()

nrow(stocknames)

write_parquet(stocknames, glue("{data_dir}/crsp-stocknames.parquet"))


# Download Compustat fundq (INTRODUCE A CHUNKED DOWNLOAD APPROACH) -------------

# Compustat quarterly fundamentals is a medium-large table. Rather than loading
# the whole thing into R memory with collect(), we use a server-side cursor to
# stream rows in batches, writing each batch to a single parquet file using
# arrow::ParquetFileWriter.

# This is the same approach used by Ian Gow's db2pqr package.
# https://github.com/iangow/db2pqr

# FIRST I DEMONSTRATE THIS APPROACH, including some overhead and technical
# details that you might not be interested in. After this first time,
# we can use the utility function download_wrds() to do this same approach but
# you won't have to see so many lines of code every time you download a file.

# How it works:
#   - dbSendQuery() creates a server-side cursor (no data transferred yet)
#   - dbFetch(n = BATCH_SIZE) pulls BATCH_SIZE rows at a time
#   - ParquetFileWriter appends each batch as a row group in the parquet file
#   - Peak memory = one batch, not the full table
#


#Some setup options
# Batch size adjusts how many rows will be batched to download at a time
# later when we wrap this in a download function we can target the batch size 
# based on the RAM available on your computer
BATCH_SIZE <- 50000

# the path is the name of the output parquet file where we want to store the 
# downloaded data 
fundq_path <- glue("{data_dir}/fundq-raw.parquet")

# AS AN EXAMPLE: We can send some raw SQL to the WRDS postgres server to apply
# some standard filters before downloading to reduce the size of the raw file. 
# Depending on your resources and needs, you might instead want to download the 
# entire raw fundq table and then filter locally. Then you can
# reuse the raw file with different criteria without re-downloading. 
fundq_sql <- "
  SELECT gvkey, datadate, fyearq, fqtr, rdq,
         conm, cusip, cik,
         saleq, ibq, epspiq, atq, cshoq, prccq, ajexq
  FROM comp.fundq
  WHERE indfmt = 'INDL' AND datafmt = 'STD'
    AND popsrc = 'D' AND consol = 'C'
    AND fyearq >= 1970
    AND rdq IS NOT NULL
"

tictoc::tic()

# Open a server-side cursor on WRDS. No rows are transferred yet — the database
# is just holding the query plan ready to stream results when we ask.
res <- dbSendQuery(wrds, fundq_sql)

# Open the output parquet file for writing (a low-level Arrow byte stream).
sink <- arrow::FileOutputStream$create(fundq_path)

# create a placeholder that we can use to point to a parquet file writer
# We'll create the ParquetFileWriter later (we need its schema first).
writer <- NULL

# Setting up counters for the progress message.
total_rows <- 0
start_time <- Sys.time()

# Loop until the cursor reports no more rows on the server side.
while (!dbHasCompleted(res)) {
  # Pull the next BATCH_SIZE rows from the server into an R data.frame.
  chunk <- dbFetch(res, n = BATCH_SIZE)
  # add a failsafe: some drivers return 0 rows before dbHasCompleted flips
  # if so — bail out.
  if (nrow(chunk) == 0) break

  # Update counters and print a one-line progress indicator (\r overwrites in place).
  total_rows <- total_rows + nrow(chunk)
  elapsed <- as.numeric(difftime(Sys.time(), start_time, units = "mins"))
  cat(sprintf("\r  %s rows | %.1f min elapsed",
              format(total_rows, big.mark = ","), elapsed))

  # Convert the R data.frame to an Arrow Table (columnar, zero-copy where possible).
  tab <- arrow::Table$create(chunk)

  # First iteration only: initialize the parquet writer using the chunk's schema.
  # We use zstd compression — good ratio, fast to read back.
  if (is.null(writer)) {
    writer <- arrow::ParquetFileWriter$create(
      schema = tab$schema,
      sink = sink,
      properties = arrow::ParquetWriterProperties$create(
        column_names = tab$schema$names,
        compression = "zstd"
      )
    )
  }

  # Append this chunk to the parquet file as a new row group.
  writer$WriteTable(tab, chunk_size = BATCH_SIZE)
}

# Finalize the parquet file (writes footer metadata — required for a valid file).
if (!is.null(writer)) writer$Close()
# Close the underlying byte stream.
sink$close()
# Release the server-side cursor.
dbClearResult(res)
# Newline after the \r progress line so later output doesn't overwrite it.
cat("\n")

tictoc::toc()

# Check the result
sprintf("Saved %s rows, %.1f MB",
        format(total_rows, big.mark = ","), file.size(fundq_path) / 1e6)


# Download CRSP daily returns (download_wrds function) -------------------------

# CRSP daily stock file (v2) is the largest table we need — about 100M rows.
# Now that you've seen the raw chunked loop above (for fundq), here we use the
# download_wrds() wrapper from utils.R. It does the same thing — chunked
# dbFetch + ParquetFileWriter — but handles the connection, batching, and
# progress output for you. It also auto-sizes batches based on max_ram_mb.
#
# RAM NOTE: The default max_ram_mb = 8000 (8 GB) controls the size of each
# batch. Actual peak RAM usage will be max_ram_mb plus ~1-2 GB baseline overhead
# (R session, loaded packages, writer buffers). If you're on a machine with
# limited RAM, set a lower value, e.g., max_ram_mb = 2000 for a 4 GB machine.
# CRSP daily only has 3 columns so RAM won't be an issue here, but for wide
# tables like comp.funda (~1000 columns) the throttle makes a big difference.
#
# CRSP v2 column names differ from the old WRDS CRSP format:
#   dsf_v2:            dlycaldt (date), dlyret (return)
#   inddlyseriesdata:  dlycaldt, dlytotret (market return), indno (index ID)
#   indno = 1000200 is the NYSE/NYSEMKT/Nasdaq/Arca VW market index

# This will take 10-15+ minutes. The progress line updates as it goes.
tictoc::tic()

download_wrds(
  sql = "SELECT permno, dlycaldt, dlyret
         FROM crsp.dsf_v2
         WHERE dlycaldt >= '1970-01-01'",
  output_path = glue("{data_dir}/crsp-dsf-v2.parquet"),
  max_ram_mb = 8000, 
  wrds_user = keyring::key_get("wrds_user"),
  wrds_pw   = keyring::key_get("wrds_pw")
)

tictoc::toc()


# --- Market index returns (inddlyseriesdata) ---

# Same download_wrds() pattern for the market index.
# the market index will be used to compute abnormal returns relative to it.
# indno = 1000200 is the NYSE/NYSEMKT/Nasdaq/Arca VW market index.
tictoc::tic()

download_wrds(sql = "SELECT dlycaldt, dlytotret
                     FROM crsp.inddlyseriesdata
                     WHERE indno = 1000200
                       AND dlycaldt >= '1970-01-01'",
              output_path = glue("{data_dir}/crsp-index.parquet"),
              wrds_user = keyring::key_get("wrds_user"),
              wrds_pw   = keyring::key_get("wrds_pw"))

tictoc::toc()


# Disconnect from WRDS ---------------------------------------------------------

dbDisconnect(wrds)

# At this point you have the following parquet files on disk:
#   {data_dir}/ccm-link.parquet         — CCM link table (gvkey -> permno)
#   {data_dir}/crsp-stocknames.parquet  — CRSP SIC codes by permno + date range
#   {data_dir}/fundq-raw.parquet        — Compustat quarterly fundamentals
#   {data_dir}/crsp-dsf-v2.parquet      — CRSP daily stock returns
#   {data_dir}/crsp-index.parquet       — CRSP daily market index returns
#
# Next: run src/2-transform-data.R to merge these tables, create variables
# (SUE, controls), and link with CRSP returns for the event study.

File: src/1-download-data.py
# ==============================================================================
# 1-download-data.py
#
# Purpose:
#   Download raw WRDS tables to local parquet files. Demonstrates two
#   download approaches at increasing levels of RAM efficiency:
#     1. polars.read_database() into a DataFrame (CCM link, CRSP
#        stocknames) — simplest, full result in memory.
#     2. Chunked pyarrow.ParquetWriter via a psycopg2 server-side cursor
#        (Compustat fundq, CRSP daily, market index). Peak memory = one
#        batch, bounded regardless of table size.
#
# Inputs:
#   WRDS PostgreSQL endpoint (credentials from the OS keyring; service
#   `wrds`, keys `username` and `password`, stored once by setup).
#
# Outputs (to RAW_DATA_DIR):
#   ccm-link.parquet         CCM link table (gvkey -> permno)
#   crsp-stocknames.parquet  CRSP SIC codes by permno + date range
#   fundq-raw.parquet        Compustat quarterly fundamentals
#   crsp-dsf-v2.parquet      CRSP daily stock returns
#   crsp-index.parquet       CRSP value-weighted market index returns
#
# Notes:
#   - Run via `uv run src/1-download-data.py`. uv manages the virtual
#     env and installs dependencies from pyproject.toml automatically.
#   - download_parquet() (the wrapper for the chunked approach) defaults
#     to skip_if_exists=True, so a re-run does NOT re-pull anything
#     already on disk. Delete a file in RAW_DATA_DIR to force a refresh.
#   - Runtime: ~15-20 minutes for the full pull. CRSP daily is the long
#     pole at ~100M rows.
# ==============================================================================


# Setup ------------------------------------------------------------------------

import os
import sys
import time
from pathlib import Path

import keyring
import polars as pl
import psycopg2

# Windows terminals default to cp1252 encoding, which can't render the Unicode
# box-drawing characters that polars uses for DataFrame display. Switching
# stdout to UTF-8 fixes this.
sys.stdout.reconfigure(encoding="utf-8")
import pyarrow as pa
import pyarrow.parquet as pq
from dotenv import load_dotenv

# Helpers from utils.py (download_parquet, batch_run, project_setup, ...).
from utils import download_parquet, project_setup

# First-time setup: prompts for paths, language combo, and WRDS credentials,
# then writes .env. Idempotent — instant no-op if .env already exists. The
# .env file's existence is the "have I been set up?" flag, no separate
# state needed.
project_setup()

# Load paths from .env (which project_setup() has just created or was
# already there). override=True ensures the .env file wins over any
# system-level environment variables with the same name.
load_dotenv(".env", override=True)
raw_data_dir = os.getenv("RAW_DATA_DIR")
data_dir = os.getenv("DATA_DIR")
output_dir = os.getenv("OUTPUT_DIR")


# Connect to WRDS --------------------------------------------------------------

# We use the keyring package to securely retrieve WRDS credentials from your
# operating system's credential store (Windows Credential Manager, macOS
# Keychain). This avoids putting passwords in .env files or in your code.
#
# FIRST TIME SETUP — run these lines in a Python console:
#   import keyring
#   keyring.set_password("wrds", "username", "your_wrds_username")
#   keyring.set_password("wrds", "password", "your_wrds_password")
# This stores credentials in your system's keychain. You only need to do
# this once per computer. To update after a password change, re-run above.

# psycopg2 is a PostgreSQL adapter for Python. WRDS exposes its data through
# a PostgreSQL server, so we connect to it directly with SQL.
# Credentials are retrieved inline from keyring so they never sit in a variable.
wrds = psycopg2.connect(
    host="wrds-pgdata.wharton.upenn.edu",
    port=9737,
    user=keyring.get_password("wrds", "username"),
    password=keyring.get_password("wrds", "password"),
    dbname="wrds",
    sslmode="require",
)

print(f"Connected to WRDS: {wrds.info.host}:{wrds.info.port}")


# Download CCM link table ------------------------------------------------------

# The CCM link table maps Compustat gvkey to CRSP permno over time.
# It's small (~87K rows) so we load it all at once with polars.read_database().
#
# polars.read_database() sends a SQL query to the database and returns the
# full result as a polars DataFrame. This is the simplest download method —
# fine when the table fits comfortably in RAM.
#
# The skip-if-exists guard around the download block means a replication run
# (or re-running this script) won't re-pull the file unless you delete it
# first. Matches the JAR data policy expectation that raw inputs are
# preserved verbatim from the analyst's original pull.

ccm_path = f"{raw_data_dir}/ccm-link.parquet"

if Path(ccm_path).exists():
    print(f"Skipping CCM link download — file exists: {ccm_path}")
else:
    start = time.time()

    ccm_link = pl.read_database(
        "SELECT * FROM crsp.ccmxpf_lnkhist",
        connection=wrds,
    )

    print(f"CCM link: {ccm_link.shape[0]:,} rows, {time.time() - start:.1f}s")
    print(ccm_link.head())

    ccm_link.write_parquet(ccm_path)


# Download CRSP stocknames_v2 --------------------------------------------------

# This table provides time-varying SIC codes by permno and date range.
# Small table, so polars.read_database() is fine.

stocknames_path = f"{raw_data_dir}/crsp-stocknames.parquet"

if Path(stocknames_path).exists():
    print(f"Skipping stocknames download — file exists: {stocknames_path}")
else:
    start = time.time()

    stocknames = pl.read_database(
        "SELECT permno, namedt, nameenddt, siccd FROM crsp.stocknames_v2",
        connection=wrds,
    )

    print(f"Stocknames: {stocknames.shape[0]:,} rows, {time.time() - start:.1f}s")

    stocknames.write_parquet(stocknames_path)


# Download Compustat fundq (CHUNKED DOWNLOAD APPROACH) -------------------------

# Compustat quarterly fundamentals is a medium-large table. Rather than loading
# the whole thing into memory, we use a server-side cursor to stream rows in
# batches, writing each batch to a single parquet file using pyarrow's
# ParquetWriter.
#
# How it works:
#   - cursor(name=...) creates a SERVER-SIDE cursor on WRDS. Passing a name
#     is what makes it server-side — without a name, psycopg2 creates a
#     client-side cursor that loads everything into memory at execute() time.
#   - cursor.fetchmany(BATCH_SIZE) pulls BATCH_SIZE rows at a time.
#   - We convert each batch to a pyarrow Table and append it to the parquet file.
#   - Peak memory = one batch, not the full table.
#
# FIRST I DEMONSTRATE THIS APPROACH with all the details visible. After this,
# we use the download_parquet() wrapper from utils.py which does the same thing
# but hides the boilerplate.

BATCH_SIZE = 50_000

# Goes to RAW_DATA_DIR — raw inputs live separately from the derived parquets
# scripts 2-4 produce.
fundq_path = f"{raw_data_dir}/fundq-raw.parquet"

# We send raw SQL to WRDS with standard Compustat filters applied server-side
# to reduce the download size. Depending on your needs, you might download
# the entire table and filter locally for reusability.
fundq_sql = """
    SELECT gvkey, datadate, fyearq, fqtr, rdq,
           conm, cusip, cik,
           saleq, ibq, epspiq, atq, cshoq, prccq, ajexq
    FROM comp.fundq
    WHERE indfmt = 'INDL' AND datafmt = 'STD'
      AND popsrc = 'D' AND consol = 'C'
      AND fyearq >= 1970
      AND rdq IS NOT NULL
"""

# Skip-if-exists guard: don't re-pull fundq if it's already on disk.
# Delete the file to force a re-download.
if Path(fundq_path).exists():
    print(f"Skipping fundq download — file exists: {fundq_path}")
else:
    start = time.time()

    # Open a server-side cursor on WRDS. No data is transferred yet — the
    # database holds the query plan ready to stream results when we ask.
    cursor = wrds.cursor(name="fundq_download")
    cursor.execute(fundq_sql)

    # We'll create the ParquetWriter on the first batch (we need its schema first).
    writer = None
    total_rows = 0

    # Loop: fetch batches until the server has no more rows.
    while True:
        # Pull the next BATCH_SIZE rows. Returns a list of tuples (one per row).
        rows = cursor.fetchmany(BATCH_SIZE)
        if not rows:
            break

        total_rows += len(rows)
        elapsed = time.time() - start
        size_mb = Path(fundq_path).stat().st_size / 1e6 if Path(fundq_path).exists() else 0
        print(
            f"\r  {total_rows:,} rows | {elapsed / 60:.1f} min | "
            f"~{size_mb:.0f} MB on disk",
            end="",
        )

        # Convert the list of tuples to a pyarrow Table.
        # cursor.description gives us the column names from the SQL result.
        col_names = [desc[0] for desc in cursor.description]
        # Transpose rows (list of tuples) into columns (dict of lists)
        col_data = {col: [row[i] for row in rows] for i, col in enumerate(col_names)}
        table = pa.Table.from_pydict(col_data)

        # Cast DECIMAL columns to float64. PostgreSQL NUMERIC precision can vary
        # between batches (e.g., decimal128(9,4) then decimal128(8,4)), which
        # causes a schema mismatch error in the ParquetWriter. Casting to float64
        # is standard for financial data and avoids this issue.
        new_schema = pa.schema([
            pa.field(f.name, pa.float64()) if pa.types.is_decimal(f.type) else f
            for f in table.schema
        ])
        table = table.cast(new_schema)

        # First batch: create the ParquetWriter using the batch's schema.
        # We use zstd compression — good ratio, fast to read back.
        if writer is None:
            writer = pq.ParquetWriter(fundq_path, table.schema, compression="zstd")

        # Append this batch to the parquet file as a new row group.
        writer.write_table(table)

    # Finalize the parquet file (writes footer metadata — required for valid file).
    if writer is not None:
        writer.close()

    # Close the server-side cursor to free resources on WRDS.
    cursor.close()

    print()  # newline after progress line
    elapsed = time.time() - start
    print(f"Saved {total_rows:,} rows, {Path(fundq_path).stat().st_size / 1e6:.1f} MB, {elapsed:.0f}s")


# Download CRSP daily returns (download_parquet wrapper) --------------------------

# CRSP daily stock file (v2) is the largest table — about 100M rows.
# Now that you've seen the raw chunked loop above (for fundq), here we use
# download_parquet() from utils.py. It does the same thing — server-side cursor
# + fetchmany + ParquetWriter — but handles the boilerplate for you.
#
# CRSP v2 column names:
#   dsf_v2:            dlycaldt (date), dlyret (return)
#   inddlyseriesdata:  dlycaldt, dlytotret (market return), indno (index ID)
#   indno = 1000200 is the NYSE/NYSEMKT/Nasdaq/Arca VW market index

print("\nDownloading CRSP daily returns (this will take 10-15+ minutes)...")

# download_parquet() defaults to skip_if_exists=True, so re-running this script
# does not re-pull tables that are already on disk.
start = time.time()

download_parquet(
    sql="SELECT permno, dlycaldt, dlyret FROM crsp.dsf_v2 WHERE dlycaldt >= '1970-01-01'",
    output_path=f"{raw_data_dir}/crsp-dsf-v2.parquet",
    connection=wrds,
)

print(f"  {time.time() - start:.0f}s total")


# Market index returns ---------------------------------------------------------

# The market index is used to compute abnormal returns (BHAR).
# indno = 1000200 is the NYSE/NYSEMKT/Nasdaq/Arca value-weighted market index.

start = time.time()

download_parquet(
    sql="""SELECT dlycaldt, dlytotret
           FROM crsp.inddlyseriesdata
           WHERE indno = 1000200
             AND dlycaldt >= '1970-01-01'""",
    output_path=f"{raw_data_dir}/crsp-index.parquet",
    connection=wrds,
)

print(f"  {time.time() - start:.0f}s total")


# Disconnect from WRDS ---------------------------------------------------------

wrds.close()

print("\nDone. Parquet files saved to:", raw_data_dir)
print("  ccm-link.parquet         — CCM link table (gvkey -> permno)")
print("  crsp-stocknames.parquet  — CRSP SIC codes by permno + date range")
print("  fundq-raw.parquet        — Compustat quarterly fundamentals")
print("  crsp-dsf-v2.parquet      — CRSP daily stock returns")
print("  crsp-index.parquet       — CRSP daily market index returns")
print("\nNext: run src/2-transform-data.py to merge and create variables.")
